File: spec-flow/src/SpecFlow.Eval/Backend.cs
namespace SpecFlow.Eval;

/// <summary>Where a tool's evaluation store lives, chosen by configuration.</summary>
/// <remarks>
/// The spelling is `eval-core`'s, so both runtimes read the same
/// <c>EVAL_STORE</c> value and a repo moves its tools together or not at all.
/// Moving from disk to object storage is an edit to configuration, never to a
/// call site.
/// </remarks>
public abstract record Backend
{
    /// <summary>The variable naming the backend.</summary>
    public const string Environment = "EVAL_STORE";

    /// <summary>Files under a directory in the working tree.</summary>
    public sealed record Disk(string Root) : Backend
    {
        /// <inheritdoc />
        public override IBlobs Open() => new DiskBlobs(Root);

        /// <inheritdoc />
        public override string Describe() => $"disk at {Root}";
    }

    /// <summary>
    /// Blobs in an Azure storage container.
    /// </summary>
    /// <remarks>
    /// Named before it is built, so adopting it is a configuration change and
    /// not a redesign. <see cref="Open"/> says plainly that it is not built
    /// rather than falling back to disk — a tool that silently writes somewhere
    /// other than where it was told is worse than one that stops.
    /// </remarks>
    public sealed record Azure(string Account, string Container, string Prefix) : Backend
    {
        /// <inheritdoc />
        public override IBlobs Open() => throw new InvalidOperationException(
            $"the azure backend is declared but not built — configured for {Account}/{Container}.\n  "
          + "The store's shape does not change with it: the same keys, the same records. "
          + $"Until it is built, set {Environment} to a path.");

        /// <inheritdoc />
        public override string Describe() =>
            Prefix.Length is 0 ? $"azure {Account}/{Container}" : $"azure {Account}/{Container}/{Prefix}";
    }

    /// <summary>Open the backend, or say why it could not be opened.</summary>
    public abstract IBlobs Open();

    /// <summary>What to call this backend in a log line.</summary>
    public abstract string Describe();

    /// <summary>Read the backend from the environment, falling back to a disk default.</summary>
    public static Backend FromEnvironment(string defaultRoot) =>
        System.Environment.GetEnvironmentVariable(Environment) is { Length: > 0 } raw
            ? Parse(raw)
            : new Disk(defaultRoot);

    /// <summary>
    /// Parse the spelling <c>EVAL_STORE</c> uses.
    /// </summary>
    /// <remarks>
    /// Either a path — taken as a disk root — or
    /// <c>azure:&lt;account&gt;/&lt;container&gt;[/&lt;prefix&gt;]</c>.
    /// A half-written azure spelling is refused rather than guessed at.
    /// </remarks>
    public static Backend Parse(string raw)
    {
        const string scheme = "azure:";
        if (!raw.StartsWith(scheme, StringComparison.Ordinal))
        {
            return new Disk(raw);
        }

        var parts = raw[scheme.Length..].Split('/', 3);
        if (parts.Length < 2 || parts[0].Length is 0 || parts[1].Length is 0)
        {
            throw new ArgumentException(
                $"`{raw}` is not a store; use a path, or azure:<account>/<container>[/<prefix>]",
                nameof(raw));
        }
        return new Azure(parts[0], parts[1], parts.Length > 2 ? parts[2] : string.Empty);
    }
}

File: eval-core/src/backend.rs
//! Which store a tool writes to, chosen by configuration rather than by code.
//!
//! The point of naming the backends in one enum is that moving from disk to
//! object storage is an edit to configuration, not to a call site. Every tool
//! resolves its store the same way, so they move together or not at all.

use std::path::PathBuf;

use product_core::error::{ProductError, Result};
use serde::{Deserialize, Serialize};

use crate::blobs::{Blobs, DiskBlobs};

/// Where a tool's evaluation store lives.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Backend {
    /// Files under a directory in the working tree.
    Disk {
        /// The store root. Relative paths resolve against the repo.
        root: PathBuf,
    },
    /// Blobs in an Azure storage container.
    ///
    /// Named here before it is built, so adopting it is a configuration change
    /// and not a redesign. [`Backend::open`] says plainly that it is not built
    /// rather than falling back to disk — a tool that silently writes somewhere
    /// other than where it was told is worse than one that stops.
    Azure {
        /// The storage account.
        account: String,
        /// The container within it.
        container: String,
        /// A key prefix inside the container, so one container holds several tools.
        #[serde(default)]
        prefix: String,
    },
}

/// The variable naming the backend, for a tool configured from the environment.
pub const BACKEND_ENV: &str = "EVAL_STORE";

impl Backend {
    /// Disk, under a repo-relative directory.
    pub fn disk(root: impl Into<PathBuf>) -> Self {
        Self::Disk { root: root.into() }
    }

    /// Read the backend from the environment, falling back to a disk default.
    ///
    /// `EVAL_STORE` is either a path — taken as a disk root — or
    /// `azure:<account>/<container>[/<prefix>]`. Anything else is refused
    /// rather than guessed at.
    pub fn from_env(default_root: impl Into<PathBuf>) -> Result<Self> {
        let Some(raw) = std::env::var_os(BACKEND_ENV) else {
            return Ok(Self::disk(default_root));
        };
        let raw = raw.to_string_lossy().into_owned();
        Self::parse(&raw)
    }

    /// Parse the spelling `EVAL_STORE` uses.
    pub fn parse(raw: &str) -> Result<Self> {
        let Some(rest) = raw.strip_prefix("azure:") else {
            return Ok(Self::disk(raw));
        };
        let mut parts = rest.splitn(3, '/');
        match (parts.next(), parts.next(), parts.next()) {
            (Some(account), Some(container), prefix)
                if !account.is_empty() && !container.is_empty() =>
            {
                Ok(Self::Azure {
                    account: account.to_string(),
                    container: container.to_string(),
                    prefix: prefix.unwrap_or_default().to_string(),
                })
            }
            _ => Err(ProductError::ConfigError(format!(
                "`{raw}` is not a store; use a path, or azure:<account>/<container>[/<prefix>]"
            ))),
        }
    }

    /// Open the backend, or say why it could not be opened.
    pub fn open(&self) -> Result<Box<dyn Blobs>> {
        match self {
            Self::Disk { root } => Ok(Box::new(DiskBlobs::new(root.clone()))),
            Self::Azure { account, container, .. } => Err(ProductError::ConfigError(format!(
                "the azure backend is declared but not built — \
                 configured for {account}/{container}.\n  \
                 The store's shape does not change with it: the same keys, the same records. \
                 Until it is built, set {BACKEND_ENV} to a path."
            ))),
        }
    }

    /// What to call this backend in a log line.
    pub fn describe(&self) -> String {
        match self {
            Self::Disk { root } => format!("disk at {}", root.display()),
            Self::Azure { account, container, prefix } if prefix.is_empty() => {
                format!("azure {account}/{container}")
            }
            Self::Azure { account, container, prefix } => {
                format!("azure {account}/{container}/{prefix}")
            }
        }
    }
}

#[path = "backend_tests.rs"]
#[cfg(test)]
mod tests;
